File: ts/src/trace.ts
import type { Context, Trace, TraceEntry } from "./types.js";

export function createTrace(nodes: TraceEntry[], ctx: Context): Trace {
  return {
    all: () => nodes,
    get: (name) => nodes.find((e) => e.name === name),
    ctx: () => ctx,
  };
}

File: py/src/rulecat/trace.py
from __future__ import annotations
from rulecat.types import Context, TraceEntry


class Trace:
    """한 Run의 읽기 전용 뷰: 전 노드 entry + ctx."""

    def __init__(self, nodes: list[TraceEntry], ctx: Context) -> None:
        self._nodes = nodes
        self._ctx = ctx

    def all(self) -> list[TraceEntry]:
        return self._nodes

    def get(self, name: str) -> TraceEntry | None:
        for e in self._nodes:
            if e.name == name:
                return e
        return None

    def ctx(self) -> Context:
        return self._ctx
